### PR TITLE
Getcwd

### PR DESCRIPTION
**Pull request by youngfalcon**
Implemented the getcwd syscall.

### CODE REVIEW

**Comment by kapkic** [comment]
Looks good

**Review by kapkic — APPROVED** [review]

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -1726,4 +1726,22 @@ impl Cage {
             syscall_error(Errno::EBADF, "getdents", "Invalid file descriptor")
         }
     }
+
+    //------------------------------------GETCWD SYSCALL------------------------------------
+    
+    pub fn getcwd_syscall(&self, buf: *mut u8, bufsize: u32) -> i32 {
+        let cwd: String = format!("{}{}", self.cwd.read().unwrap().to_string_lossy(), "\0");
+        let bytes: Vec<u8> = cwd.as_bytes().to_vec();
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1734]: I don't think this is a very good implementation. We'd prefer to_string to to_string_lossy. Then you'd just have an &str which you can use as_bytes on directly. without needing to format anything. To add the terminating null byte, just push a 0 onto the stack.

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -1726,4 +1726,22 @@ impl Cage {
...
+        let length = bytes.len();
+
+        use std::convert::TryInto;
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1737]: We try very hard to import as little as possible even from the rust standard library and when we do we make sure to do it through the interface.

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -1726,4 +1726,22 @@ impl Cage {
...
+        if bufsize < length.try_into().unwrap() {
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1738]: to solve this problem all you need to do is do `bufsize as usize`, no need for any fancy try_into

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -1726,4 +1726,22 @@ impl Cage {
...
+            return syscall_error(Errno::ERANGE, "getcwd", "the length (in bytes) of the absolute pathname of the current working directory exceeds the given size");
+        }
+        
+        let slice = unsafe{std::slice::from_raw_parts_mut(buf, length)};
```
Comment (jesings) [inline src/safeposix/syscalls/fs_calls.rs:1742]: We additionally don't use any unsafe code except for in the interface, so this should probably be handled into the dispatcher. I'd say that this function should probably take in an `&mut [u8]`, which you can get from `get_slice_from_string`, which you should use in the dispatcher. Check the implementation of Epoll for something similar we did, although the slice was not of u8s, it should be relatively easy to understand how to transform these things.

File: src/tests/fs_tests.rs [hunk 5]
```diff
@@ -868,7 +869,8 @@ mod fs_tests {
         
         assert_eq!(cage.mkdir_syscall("/getdents", S_IRWXA), 0);
         let fd = cage.open_syscall("/getdents", O_RDWR, S_IRWXA);
-        assert_eq!(cage.getdents_syscall(fd, baseptr, bufsize), 48);
+        use std::convert::TryInto;
+        assert_eq!(cage.getdents_syscall(fd, baseptr, bufsize.try_into().unwrap()), 48);
```
Comment (jesings) [inline src/tests/fs_tests.rs:873]: Rather than use try_into, use an explicit cast

File: src/tests/fs_tests.rs [hunk 6]
```diff
@@ -889,4 +891,34 @@ mod fs_tests {
...
         lindrustfinalize();
     }
+
+    pub fn ut_lind_fs_dir_chdir_getcwd() {
+        lindrustinit();
+        let cage = {CAGE_TABLE.read().unwrap().get(&1).unwrap().clone()};
+        let needed = "/subdir1\0".as_bytes().to_vec().len();
+
+        use std::convert::TryInto;
+        let needed_u32: u32 = needed.try_into().unwrap();
```
Comment (jesings) [inline src/tests/fs_tests.rs:901]: Again here just use `as u32`

File: src/tests/fs_tests.rs [hunk 7]
```diff
@@ -889,4 +891,34 @@ mod fs_tests {
...
+        assert_eq!(String::from_utf8_lossy(&buf), "/\0\0\0\0\0\0\0\0");
+
+        cage.mkdir_syscall("/subdir1", S_IRWXA);
+        assert_eq!(cage.access_syscall("subdir1", F_OK), 0);
+        assert_eq!(cage.chdir_syscall("subdir1"), 0);
+
+        assert_eq!(cage.getcwd_syscall(bufptr, 0), -(Errno::ERANGE as i32));
+        assert_eq!(cage.getcwd_syscall(bufptr, needed_u32-1), -(Errno::ERANGE as i32));
+        assert_eq!(cage.getcwd_syscall(bufptr, needed_u32), 0);
+        assert_eq!(String::from_utf8_lossy(&buf), "/subdir1\0");
```
Comment (jesings) [inline src/tests/fs_tests.rs:919]: You can/should probably use the non lossy versions of these.

File: src/safeposix/syscalls/fs_calls.rs (hunk 4 above)
Comment (jesings) [reply src/safeposix/syscalls/fs_calls.rs:1742]: The epoll implementations are in the sockremoval_better branch

**Review by jesings — COMMENTED** [review]
Overall a very impressive first contribution to rustposix. There are a few changes that should be made before this is merge ready, which I've commented on above.

File: src/tests/fs_tests.rs (hunk 7 above)
Comment (youngfalcon) [reply src/tests/fs_tests.rs:919]: Why do we not want the lossy versions? Is a panic preferable to an un-exact string?

File: src/safeposix/syscalls/fs_calls.rs (hunk 2 above)
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: Fixed with the last update.

File: src/safeposix/syscalls/fs_calls.rs (hunk 3 above)
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: Fixed with the last update.

File: src/tests/fs_tests.rs (hunk 6 above)
Comment (youngfalcon) [reply src/tests/fs_tests.rs]: Fixed with the last update.

File: src/tests/fs_tests.rs (hunk 5 above)
Comment (youngfalcon) [reply src/tests/fs_tests.rs]: Fixed with the last update.

File: src/tests/fs_tests.rs (hunk 7 above)
Comment (jesings) [reply src/tests/fs_tests.rs:917]: In the tests we know that you're passing/getting passed valid utf8, so it won't return an error variant (which I assume is what you mean by panic as that would panic on an unwrap). Additionally, the lossy and non-lossy versions actually do somewhat different things--the non-lossy version allocates a new buffer whereas the lossy version does it in place and wraps it within a Cow<> object which makes it copy-on-write. This means it's somewhat more complex and annoying to deal with, and we don't care about the overhead of the utf8 validity check within the test.

Comment (youngfalcon) [reply src/tests/fs_tests.rs]: Updated this to use std::str::from_utf8.

(String::from_utf8 does not work directly as it requires the ownership of the passed vector.)

**Comment by rennergade** [comment]
I've added a function to the interface to convert mut buf ptr's with sizes to mut slices. You can merge from the branch `addmutslice` to gain that function here.

File: src/safeposix/syscalls/fs_calls.rs (hunk 4 above)
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: Updated this with the last few commits.

File: src/safeposix/syscalls/fs_calls.rs (hunk 1 above)
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: Updated.

File: src/safeposix/syscalls/fs_calls.rs [hunk 8]
```diff
@@ -1722,4 +1722,20 @@ impl Cage {
             syscall_error(Errno::EBADF, "getdents", "Invalid file descriptor")
         }
     }
+
+    //------------------------------------GETCWD SYSCALL------------------------------------
+    
+    pub fn getcwd_syscall(&self, buf: *mut u8, bufsize: u32) -> i32 {
+        let mut bytes: Vec<u8> = self.cwd.read().unwrap().to_str().unwrap().as_bytes().to_vec();
+        bytes.push(0u8);
```
Comment (rennergade) [inline src/safeposix/syscalls/fs_calls.rs:1730]: Being pedantic but would love a short comment here about null terminating.

**Review by rennergade — APPROVED** [review]
This looks really great now. Succinct and functional.

Just add that one comment and I'll merge it.

File: src/safeposix/syscalls/fs_calls.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: This issue was fixed in previous commits.

File: src/safeposix/syscalls/fs_calls.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: This issue was fixed in previous commits.

File: src/safeposix/syscalls/fs_calls.rs (hunk 8 above)
Comment (youngfalcon) [reply src/safeposix/syscalls/fs_calls.rs]: Updated.

**Pull request merged**